tensorflow_quantum/core/ops/noise/tfq_noisy_samples.cc

```diff
@@ -0,0 +1,348 @@
+        sv = ss.Create(largest_nq);
+        scratch = ss.Create(largest_nq);
+      }
+
+      QTSimulator::Parameter param;
+      param.collect_kop_stat = false;
+      param.collect_mea_stat = true;
+      param.normalize_before_mea_gates = true;
+
+      // Track op-wise stats.
+      std::vector<uint64_t> gathered_samples;
+
+      for (int j = 0; j < num_samples; j++) {
+        ss.SetStateZero(sv);
+        // time since epoch seeds random generator.
+        unsigned long r_seed =
+            std::chrono::duration_cast<std::chrono::milliseconds>(
+                std::chrono::system_clock::now().time_since_epoch())
+                .count();
+
+        QTSimulator::RunOnce(param, ncircuits[i], r_seed, ss, sim, scratch, sv,
+                             gathered_samples);
+        uint64_t q_ind = 0;
+        uint64_t mask = 1;
+        bool val = 0;
+        while (q_ind < nq) {
+          val = gathered_samples[0] & mask;
+          (*output_tensor)(
+              i, j, static_cast<ptrdiff_t>(max_num_qubits - q_ind - 1)) = val;
+          q_ind++;
+          mask <<= 1;
+        }
+        while (q_ind < max_num_qubits) {
+          (*output_tensor)(
+              i, j, static_cast<ptrdiff_t>(max_num_qubits - q_ind - 1)) = -2;
+          q_ind++;
+        }
+      }
+    }
+  }
+
+  void ComputeSmall(const std::vector<int>& num_qubits,
+                    const int max_num_qubits, const int num_samples,
+                    const std::vector<NoisyQsimCircuit>& ncircuits,
+                    tensorflow::OpKernelContext* context,
+                    tensorflow::TTypes<int8_t, 3>::Tensor* output_tensor) {
+    using Simulator = qsim::Simulator<const qsim::SequentialFor&>;
+    using StateSpace = Simulator::StateSpace;
+    using QTSimulator =
+        qsim::QuantumTrajectorySimulator<qsim::IO, QsimGate,
+                                         qsim::MultiQubitGateFuser, Simulator>;
+
+    const int output_dim_batch_size = output_tensor->dimension(0);
+    const int num_threads = context->device()
+                                ->tensorflow_cpu_worker_threads()
+                                ->workers->NumThreads();
+
+    // [num_threads, batch_size].
+    std::vector<std::vector<int>> rep_offsets(
+        num_threads, std::vector<int>(output_dim_batch_size, 0));
+    BalanceTrajectory(num_samples, num_threads, &rep_offsets);
+
+    // [num_threads, batch_size] stores the number of
+    // samples written by thread range [0, i].
+    std::vector<std::vector<long>> offset_prefix_sum(
+        num_threads, std::vector<long>(output_dim_batch_size, 0));
+
+    for (int i = 0; i < output_dim_batch_size; i++) {
+      int p_reps = (num_samples + num_threads - 1) / num_threads;
+      offset_prefix_sum[0][i] = rep_offsets[0][i] + p_reps;
+      for (int j = 1; j < num_threads; j++) {
+        offset_prefix_sum[j][i] += offset_prefix_sum[j - 1][i];
+        offset_prefix_sum[j][i] += rep_offsets[j][i] + p_reps;
```

Comment:
I don't understand why `offset_prefix_sum[j][i]` varies over `batch_size`. Since we have the same `num_samples` for all batch, can we just use `offset_prefix_sum[j]`?

Reply:
This is a little tricky. `offset_pref_sum[j][i]` stores the `final sample number i that thread j should draw` If I do `offset_pref_sum[j][i] - offset_prefix_sum[j-1][i]` I get `the number of samples that thread j should draw from circuit i`. And finally `offset_prefix_sum[j-1][i]` tells `thread j where to start storing sample i` that it draws from the circuit. This allows one to do non locking writes to the output tensor since each thread knows exactly which indices in the output tensor it needs to write to.

`BalanceTrajectory` divides up the work across the threads evenly so in the event of: `n_samples = 10`, `n_circuits = 9` and `num_threads=11` The BalanceTrajectory function will assign offsets so that threads [1,10] all draw 1 sample from the first circuit with the 11th thread drawing no samples from the first circuit. To balance this out the 11th thread would then draw two samples from the second circuit and so on.... so the values of offset_prefix_sum[j][i] might be very close to one another, but they do have small offsets to make sure things are balanced and we can't just use `offset_prefix_sum[j]`. 

`offset_prefix_sum` stores all the information for the threads to do this balancing between them.

Does this clear things up ?

Reply:
Two thumbs up!